gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(5);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(6) ? internalRow.getString(6) : null;
+    boolean dryRun = !internalRow.isNullAt(7) && internalRow.getBoolean(7);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    runId = updateRunId(gcOutputTableIdentifier, runId, identifiedResultsRepo);
+
+    Map<String, List<Long>> expiredSnapshotsPerContentId =
+        getExpiredSnapshotsPerContentId(runId, identifiedResultsRepo);
+
+    Map<String, List<String>> metadataLocationsPerContentId =
+        getMetadataLocationsPerContentId(runId, identifiedResultsRepo);
+
+    List<InternalRow> outputRows = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(nessieClientConfig)) {
+      createGCBranch(expiryBranchName, api);
+      // get Nessie catalog
+      Catalog nessieCatalog = GCUtil.loadNessieCatalog(spark(), gcCatalogName, expiryBranchName);
+      String finalRunId = runId;
+      expiredSnapshotsPerContentId
+          .entrySet()
+          .forEach(
+              entry -> {
+                // Expiry Algorithm:
+                // A. For each content id, fetch ANY one of the content.
+                // B. Use the newly created expiry branch. On this branch,
+                // C. Do a single PUT operation with a fetched content from step-A and
+                //    a dummy ContentKey.
+                // D. Collect the list of all the expired snapshot for this content id.
+                // E. Fetch this table using dummy ContentKey on expiry branch. For this table,
+                // F. call expire_snapshot action from Iceberg by passing the list of expired
+                //    snapshots.
+                // G. Once all the content ids are handled, drop this branch.
+                //    (Else it may impact the next GC operations as some contents are still referred
+                // in
+                //    this branch)
+                String contentId = entry.getKey();
+                addOneContentToExpireProcedureBranch(
+                    expiryBranchName, identifiedResultsRepo, api, finalRunId, contentId);
+                Table table =
+                    nessieCatalog.loadTable(
+                        TableIdentifier.parse(EXPIRE_NAMESPACE + "." + contentId));
+
+                // TODO: Remove after https://github.com/apache/iceberg/pull/4509
+                updateMetadataLocation((BaseTable) table);
+
+                Set<Snapshot> snapshotSet =
+                    new TreeSet<>((o1, o2) -> (int) (o1.timestampMillis() - o2.timestampMillis()));
```

Comment:
This will be wrong when two snapshots have the same timestamp.

Reply:
same timestamp will be grouped together. I don't see a problem here.

Reply:
Contract of `Comparator`?
BTW: Subtraction here esp. in combination with a cast to `int` can go sideways.